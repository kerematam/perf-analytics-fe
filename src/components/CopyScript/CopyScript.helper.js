export const script = (
  pageUrl,
  pageId,
  requestUrl = 'http://localhost:4040'
) => `<!--
* Page URL : ${pageUrl}
* Page ID : ${pageId}
* Analystics Host : ${requestUrl}
-->
<script async>
function calcualteTimer() {
  var timer = {};
  var timing = window.performance.timing;
  timer.ttfb = timing.responseStart - timing.requestStart;
  timer.fcp = performance.getEntriesByType("paint")[1].startTime;
  timer.doml = timing.domComplete - timing.domLoading;
  timer.wl = timing.loadEventEnd - timing.navigationStart;
  timer.resources = performance.getEntriesByType("resource");
  return timer;
}

function sendTimer() {
  var timer = calcualteTimer()
  var body = JSON.stringify(timer);
  fetch("${requestUrl}/api/sites/${pageId}/metrics", {
      body,
      headers: {
        "Content-Type": "application/json",
      },
      method: "POST",
    }
  )
}

window.addEventListener("load", function () {
  setTimeout(sendTimer, 0);
});
</script>`

export const isValidUrl = url =>
  /((([A-Za-z]{3,9}:(?:\/\/)?)(?:[-;:&=+$,\w]+@)?[A-Za-z0-9.-]+(:[0-9]+)?|(?:www.|[-;:&=+$,\w]+@)[A-Za-z0-9.-]+)((?:\/[+~%/.\w-_]*)?\??(?:[-+=&;%@.\w_]*)#?(?:[\w]*))?)/.test(
    url
  )
