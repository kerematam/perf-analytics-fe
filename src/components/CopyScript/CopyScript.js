import React, { useState, useEffect, useMemo } from 'react'
import IconButton from '@material-ui/core/IconButton'
import TextField from '@material-ui/core/TextField'
import { FileCopy, AddBox as AddBoxIcon } from '@material-ui/icons'
import CircularProgress from '@material-ui/core/CircularProgress'
import { script, isValidUrl } from './CopyScript.helper.js'

import styles from './CopyScript.module.scss'

export const AddUrl = ({ onSubmit = () => {}, url: urlProp = '', loading }) => {
  const [url, setUrl] = useState('')
  const [isError, setIsError] = useState('')

  const handleChange = e => {
    setUrl(e.target.value)
    setIsError('')
  }

  const handleAddUrl = () => {
    const isValid = isValidUrl(url)

    if (isValid) {
      onSubmit(url)
      setIsError('')
    } else {
      setIsError('Please enter valid URL. Eg.: http://localhost:3000/test.html')
    }
  }

  useEffect(() => {
    setUrl(urlProp)
  }, [urlProp])

  const onError = !!isError ? ` ${styles.on_error}` : ''

  return (
    <div className={`${styles.add_url_container}${onError}`}>
      <TextField
        value={url}
        disabled={loading}
        onChange={handleChange}
        size="small"
        className={styles.input}
        id="outlined-basic"
        label="URL"
        variant="outlined"
        error={!!isError}
        helperText={isError}
      />
      <IconButton onClick={handleAddUrl}>
        {loading ? (
          <CircularProgress className={styles.button_icon} />
        ) : (
          <AddBoxIcon className={styles.button_icon} />
        )}
      </IconButton>
    </div>
  )
}

const CopyScript = ({ onSubmit, page = {}, reqUrl, loading }) => {
  const scriptStr = useMemo(() => script(page.url, page._id, reqUrl), [
    page,
    reqUrl,
  ])

  const handleCopy = () => {
    window.navigator.clipboard.writeText(scriptStr)
  }

  return (
    <div>
      <AddUrl url={page?.url} onSubmit={onSubmit} loading={loading} />
      <div className={styles.copy_block_container}>
        <IconButton onClick={handleCopy} className={styles.copy_icon}>
          <FileCopy />
        </IconButton>
        <TextField
          className={styles.copy_field}
          value={scriptStr}
          id="outlined-multiline-static"
          label="Copy Block"
          multiline
          rows={10}
          variant="outlined"
        />
      </div>
    </div>
  )
}

export default CopyScript
