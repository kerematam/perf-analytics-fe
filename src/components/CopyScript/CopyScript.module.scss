@import '../../themes/colors';

.add_url_container {
  display: flex;
  padding-bottom: 5px;
  .input {
    width: 100%;
    position: relative;
    top: 12px;
  }
  .button_icon {
    font-size: 40px;
  }
  transition: padding-bottom 0.3s;
}

.on_error {
  padding-bottom: 25px;
}

$copy_icon_padding: 32px;
.copy_block_container {
  position: relative;
  display: inline-block;
  width: 100%;
  .copy_icon {
    z-index: 1;
    position: absolute;
    top: 8px;
    right: $copy_icon_padding;
    &:hover {
      cursor: pointer;
      color: $PrimaryColorMain;
    }
  }
  .copy_field {
    width: 100%;
    textarea {
      padding-right: calc(10px + #{$copy_icon_padding});
      height: 300px;
      width: 100%;
      font-family: monospace;
    }
  }
}
