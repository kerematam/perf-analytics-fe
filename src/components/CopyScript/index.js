import CopyScript from './CopyScript'

export default CopyScript
