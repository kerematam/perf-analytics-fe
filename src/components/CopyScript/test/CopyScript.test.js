import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import CopyScript from '../index'
import { AddUrl } from '../CopyScript'

const setup = setupProps => {
  const wrapper = renderer.create(<CopyScript {...setupProps} />).toJSON()

  return {
    wrapper,
  }
}

const generateData = () => {
  return { page: { url: 'asd' } }
}

describe('Chart component', () => {
  test('should match with snapshot', () => {
    const mockProps = generateData()
    const { wrapper } = setup(mockProps)

    expect(wrapper).toMatchSnapshot()
  })

  test('should render correctly', () => {
    render(<CopyScript {...generateData()} />)

    const [input] = document.getElementsByTagName('input')
    expect(input.value).toContain('asd')

    const [textarea] = document.getElementsByTagName('textarea')
    expect(textarea.value).toContain('Page URL')
  })
})

describe('AddUrl component', () => {
  const INVALID_URL = '::asd'
  const VALID_URL = 'http://localhost:3000/test.html'

  test('should handle error correctly', () => {
    render(<AddUrl url={INVALID_URL} />)

    const [input] = document.getElementsByTagName('input')
    const [button] = document.getElementsByTagName('button')
    fireEvent.click(button)
    const warningMsg = document.getElementById('add_url_text_box-helper-text')
      .innerHTML

    expect(warningMsg).toContain('Please')
  })

  test('should not show error if input correct', () => {
    render(<AddUrl url={VALID_URL} />)
    const [button] = document.getElementsByTagName('button')
    fireEvent.click(button)
    const warningMsg = document.getElementById('add_url_text_box-helper-text')
      ?.innerHTML

    expect(warningMsg).toBeFalsy()
  })
})
