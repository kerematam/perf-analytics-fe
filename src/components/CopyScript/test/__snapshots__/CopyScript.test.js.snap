// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component should match with snapshot 1`] = `
<div>
  <div
    className="add_url_container"
  >
    <div
      className="MuiFormControl-root MuiTextField-root input"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="outlined-basic"
        id="outlined-basic-label"
      >
        URL
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          disabled={false}
          id="outlined-basic"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-61"
          >
            <span>
              URL
            </span>
          </legend>
        </fieldset>
      </div>
      
    </div>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root button_icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="copy_block_container"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root copy_icon"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4l6 6v10c0 1.1-.9 2-2 2H7.99C6.89 23 6 22.1 6 21l.01-14c0-1.1.89-2 1.99-2h7zm-1 7h5.5L14 6.5V12z"
          />
        </svg>
      </span>
    </button>
    <div
      className="MuiFormControl-root MuiTextField-root copy_field"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink={true}
        htmlFor="outlined-multiline-static"
        id="outlined-multiline-static-label"
      >
        Copy Block
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
        onClick={[Function]}
      >
        <textarea
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          disabled={false}
          id="outlined-multiline-static"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows={10}
          value="<!--
* Page URL : asd
* Page ID : undefined
* Analystics Host : http://localhost:4040
-->
<script async>
function calcualteTimer() {
  var timer = {};
  var timing = window.performance.timing;
  timer.ttfb = timing.responseStart - timing.requestStart;
  timer.fcp = performance.getEntriesByType(\\"paint\\")[1].startTime;
  timer.doml = timing.domComplete - timing.domLoading;
  timer.wl = timing.loadEventEnd - timing.navigationStart;
  timer.resources = performance.getEntriesByType(\\"resource\\");
  return timer;
}

function sendTimer() {
  var timer = calcualteTimer()
  var body = JSON.stringify(timer);
  fetch(\\"http://localhost:4040/api/sites/undefined/metrics\\", {
      body,
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
      method: \\"POST\\",
    }
  )
}

window.addEventListener(\\"load\\", function () {
  setTimeout(sendTimer, 0);
});
</script>"
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-61 PrivateNotchedOutline-legendNotched-62"
          >
            <span>
              Copy Block
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
